crates/ruff/src/rules/pylint/rules/redefined_loop_name.rs

```diff
@@ -0,0 +1,194 @@
+    /// assignment.
+    ///
+    /// ## Why is this bad?
+    /// Redefinition of a loop variable inside the loop's body causes its value to differ from
+    /// the original loop iteration for the remainder of the block, in a way that will likely
+    /// cause bugs.
+    ///
+    /// In Python, unlike many other languages, `for` loops and `with` statements don't define
+    /// their own scopes. Therefore, a nested loop that uses the same target variable name as
+    /// an outer loop will reuse the same actual variable, and the value from the last
+    /// iteration will "leak out" into the remainder of the enclosing loop.
+    ///
+    /// While this mistake is easy to spot in small examples, it can be hidden in larger
+    /// blocks of code where the definition and redefinition of the variable may not be
+    /// visible at the same time.
+    ///
+    /// ## Example
+    /// ```python
+    /// for i in range(10):
+    ///     i = 9
+    ///     print(i)  # prints 9 every iteration
+    ///
+    /// for i in range(10):
+    ///     for i in range(10):  # original value overwritten
+    ///         pass
+    ///     print(i)  # also prints 9 every iteration
+    ///
+    /// with path1.open() as f:
+    ///     with path2.open() as f:
+    ///         f = path2.open()
+    ///     print(f.readline())  # prints a line from path2
+    /// ```
+    pub struct RedefinedLoopName {
+        pub name: String,
+    }
+);
+impl Violation for RedefinedLoopName {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let RedefinedLoopName { name } = self;
+        format!("For loop or with statement variable `{name}` overwritten in body")
+    }
+}
+
+struct InnerForWithAssignNamesVisitor<'a> {
+    locator: &'a Locator<'a>,
+    name_ranges: Vec<Range>,
+}
+
+impl<'a, 'b> Visitor<'b> for InnerForWithAssignNamesVisitor<'_>
+where
+    'b: 'a,
+{
+    fn visit_stmt(&mut self, stmt: &'b Stmt) {
+        match &stmt.node {
+            // For and async for.
+            StmtKind::For { target, .. } | StmtKind::AsyncFor { target, .. } => {
+                self.name_ranges
+                    .extend(name_ranges_from_expr(target, self.locator));
+            }
+            // With.
+            StmtKind::With { items, .. } => {
+                self.name_ranges
+                    .extend(name_ranges_from_with_items(items, self.locator));
+            }
+            // Assignment, augmented assignment, and annotated assignment.
+            StmtKind::Assign { targets, .. } => {
+                self.name_ranges
+                    .extend(name_ranges_from_assign_targets(targets, self.locator));
+            }
+            StmtKind::AugAssign { target, .. } | StmtKind::AnnAssign { target, .. } => {
+                self.name_ranges
+                    .extend(name_ranges_from_expr(target, self.locator));
+            }
+            _ => {}
+        }
+        visitor::walk_stmt(self, stmt);
```

Comment:
Do we need to avoid recursing in any cases? E.g., what if we enter a function or class definition?

Reply:
Great point! Added exclusions for function and class definitions since they have their own scopes, after confirming that is the correct thing to do by running some quick examples.